rhp/v3/transport.go

```diff
@@ -48,43 +54,50 @@ type protocolObject interface {
 	encoding.SiaUnmarshaler
 }
 
-func writeRequest(s *mux.Stream, id Specifier, req protocolObject) error {
-	if _, err := s.Write(id[:]); err != nil {
+func writeRequest(w io.Writer, id Specifier, req protocolObject) error {
+	if _, err := w.Write(id[:]); err != nil {
 		return err
 	}
-	return req.MarshalSia(s)
+	return req.MarshalSia(w)
 }
 
-func readResponse(s *mux.Stream, resp protocolObject) error {
+func readResponse(r io.Reader, resp protocolObject) error {
 	rr := rpcResponse{nil, resp}
-	if err := rr.UnmarshalSia(s); err != nil {
+	if err := rr.UnmarshalSia(r); err != nil {
 		return err
 	} else if rr.err != nil {
 		return rr.err
 	}
 	return nil
 }
 
-func writeResponse(s *mux.Stream, resp protocolObject) error {
-	return (&rpcResponse{nil, resp}).MarshalSia(s)
+func writeResponse(w io.Writer, resp protocolObject) error {
+	return (&rpcResponse{nil, resp}).MarshalSia(w)
 }
 
 func writeResponseErr(s *mux.Stream, err error) error {
 	return (&rpcResponse{&RPCError{Description: err.Error()}, nil}).MarshalSia(s)
 }
 
-func processPayment(s *mux.Stream, payment PaymentMethod) error {
-	if err := writeResponse(s, paymentType(payment)); err != nil {
-		return err
-	} else if err := writeResponse(s, payment); err != nil {
+func processPayment(rw io.ReadWriter, payment PaymentMethod) error {
```

Comment:
@lukechampine this change and others in this PR which are not great (since we are using modules methods) were necessary since the code in `rhp/v3/encoding.go` doesn't seem to produce the same encoding as siad for rhp3. So I used this as a quick workaround since I thought you might want to check the encoding out yourself.

Reply:
hmm. Might be because `RPCWrite` calls `WriteObject`, which first encodes the object to a `[]byte` and then length-prefixes that `[]byte`.

Reply:
There's encoding for all RHP3 types in https://github.com/SiaFoundation/hostd/blob/master/rhp/v3/encoding.go if you want to move away from `siad/modules`

Reply:
Fixed it for now by updating the marshaling methods for the individual types. Once we have a common encoding package it might be worth refactoring renterd and hostd to use that instead.